File: 6.actions/src/action_py_pkg/action_py_pkg/turtle_action_server.py
#!/usr/bin/env python3
import math
import threading

import rclpy
from geometry_msgs.msg import Twist, TwistStamped
from nav_msgs.msg import Odometry
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node
from turtlesim.msg import Pose

from custom_interfaces.action import MoveToGoal
import os


class TurtleActionServer(Node):
    """Action server that drives turtle1/TurtleBot3 to a requested (target_x, target_y) coordinate."""

    def __init__(self):
        super().__init__("turtle_action_server")

        self.cb_group = ReentrantCallbackGroup()

        # Coordinate boundary configuration (relaxed by default for open Gazebo worlds)
        self.declare_parameter("enforce_canvas_bounds", False)
        self.declare_parameter("min_x", -20.0)
        self.declare_parameter("max_x", 20.0)
        self.declare_parameter("min_y", -20.0)
        self.declare_parameter("max_y", 20.0)

        # Velocity message type: ROS 2 Jazzy uses TwistStamped on /cmd_vel for Gazebo Harmonic
        default_stamped = os.environ.get("ROS_DISTRO", "jazzy") != "humble"
        self.declare_parameter("use_stamped_vel", default_stamped)
        self.use_stamped_vel = self.get_parameter("use_stamped_vel").get_parameter_value().bool_value

        # Goal tracking for preemption
        self.active_goal_handle = None
        self.goal_lock = threading.Lock()

        # Turtle / Robot state
        self.current_pose = None

        # Velocity publishers:
        # - /turtle1/cmd_vel: always geometry_msgs/msg/Twist for Turtlesim
        # - /cmd_vel: TwistStamped for ROS 2 Jazzy Gazebo Harmonic bridge, or Twist for Humble
        self.cmd_vel_pub = self.create_publisher(Twist, "/turtle1/cmd_vel", 10)
        if self.use_stamped_vel:
            self.tb3_cmd_vel_pub = self.create_publisher(TwistStamped, "/cmd_vel", 10)
        else:
            self.tb3_cmd_vel_pub = self.create_publisher(Twist, "/cmd_vel", 10)

        # Pose / Odometry subscribers (dual support: Turtlesim Pose & Gazebo Odometry)
        self.pose_sub = self.create_subscription(
            Pose,
            "/turtle1/pose",
            self.handle_pose,
            10,
            callback_group=self.cb_group,
        )
        self.odom_sub = self.create_subscription(
            Odometry,
            "/odom",
            self.handle_odom,
            10,
            callback_group=self.cb_group,
        )

        # Action server
        self.action_server = ActionServer(
            self,
            MoveToGoal,
            "move_to_goal",
            execute_callback=self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
            callback_group=self.cb_group,
        )

        self.get_logger().info("Turtle Action Server has been started! Action: /move_to_goal")

    def handle_pose(self, msg: Pose):
        """Update the latest known pose from Turtlesim."""
        self.current_pose = msg

    def handle_odom(self, msg: Odometry):
        """Update the latest known pose from Gazebo / TurtleBot3 Odometry."""
        x = msg.pose.pose.position.x
        y = msg.pose.pose.position.y
        q = msg.pose.pose.orientation
        siny_cosp = 2.0 * (q.w * q.z + q.x * q.y)
        cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
        theta = math.atan2(siny_cosp, cosy_cosp)
        self.current_pose = Pose(x=float(x), y=float(y), theta=float(theta))

    def goal_callback(self, goal_request: MoveToGoal.Goal):
        """Validate the requested target coordinates within acceptable boundaries."""
        self.get_logger().info(
            f"Received goal request: target=({goal_request.target_x:.2f}, {goal_request.target_y:.2f}), "
            f"velocity={goal_request.linear_velocity:.2f}"
        )

        enforce_bounds = self.get_parameter("enforce_canvas_bounds").get_parameter_value().bool_value
        min_x = 0.0 if enforce_bounds else self.get_parameter("min_x").get_parameter_value().double_value
        max_x = 11.0 if enforce_bounds else self.get_parameter("max_x").get_parameter_value().double_value
        min_y = 0.0 if enforce_bounds else self.get_parameter("min_y").get_parameter_value().double_value
        max_y = 11.0 if enforce_bounds else self.get_parameter("max_y").get_parameter_value().double_value

        if min_x <= goal_request.target_x <= max_x and min_y <= goal_request.target_y <= max_y:
            self.get_logger().info("Target within boundary. Accepting goal.")
            return GoalResponse.ACCEPT
        else:
            self.get_logger().warn(
                f"Rejecting goal: target ({goal_request.target_x:.2f}, {goal_request.target_y:.2f}) "
                f"is outside boundary [{min_x:.1f}, {max_x:.1f}]!"
            )
            return GoalResponse.REJECT

    def cancel_callback(self, goal_handle):
        """Accept goal cancellation requests from clients."""
        self.get_logger().info("Received goal cancellation request.")
        return CancelResponse.ACCEPT

    def stop_turtle(self):
        """Halt robot motion by publishing zero velocity to both Turtlesim and Gazebo."""
        stop_cmd = Twist()
        self.cmd_vel_pub.publish(stop_cmd)
        if self.use_stamped_vel:
            stop_stamped = TwistStamped()
            stop_stamped.header.stamp = self.get_clock().now().to_msg()
            stop_stamped.header.frame_id = "base_footprint"
            self.tb3_cmd_vel_pub.publish(stop_stamped)
        else:
            self.tb3_cmd_vel_pub.publish(stop_cmd)

    def execute_callback(self, goal_handle):
        """Execute the motion toward the target, publishing feedback and handling cancellation."""
        target_x = goal_handle.request.target_x
        target_y = goal_handle.request.target_y
        linear_velocity = (
            goal_handle.request.linear_velocity
            if goal_handle.request.linear_velocity > 0.0
            else 2.0
        )

        with self.goal_lock:
            if self.active_goal_handle is not None and self.active_goal_handle.is_active:
                self.get_logger().info("Preempting previous active goal for new target...")
            self.active_goal_handle = goal_handle

        feedback_msg = MoveToGoal.Feedback()
        result = MoveToGoal.Result()

        loop_rate = self.create_rate(10)
        start_time = self.get_clock().now()
        total_distance = 0.0
        prev_pose = self.current_pose

        self.get_logger().info(f"Executing goal: moving to ({target_x:.2f}, {target_y:.2f})...")

        while rclpy.ok():
            # 1. Check if preempted by a newer goal request
            with self.goal_lock:
                if goal_handle != self.active_goal_handle:
                    self.get_logger().info(
                        f"Goal to ({target_x:.2f}, {target_y:.2f}) was preempted by a new goal."
                    )
                    goal_handle.abort()
                    result.success = False
                    result.total_distance = total_distance
                    result.elapsed_time = (self.get_clock().now() - start_time).nanoseconds / 1e9
                    return result

            # 2. Check if cancellation was requested
            if goal_handle.is_cancel_requested:
                self.stop_turtle()
                goal_handle.canceled()
                self.get_logger().info("Goal canceled by client.")
                with self.goal_lock:
                    if self.active_goal_handle == goal_handle:
                        self.active_goal_handle = None
                result.success = False
                result.total_distance = total_distance
                result.elapsed_time = (self.get_clock().now() - start_time).nanoseconds / 1e9
                return result

            # Wait until initial pose is available
            if self.current_pose is None:
                loop_rate.sleep()
                continue

            # 3. Accumulate traveled distance
            if prev_pose is not None:
                total_distance += math.hypot(
                    self.current_pose.x - prev_pose.x,
                    self.current_pose.y - prev_pose.y,
                )
            prev_pose = self.current_pose

            # 4. Calculate distance and angle to target
            delta_x = target_x - self.current_pose.x
            delta_y = target_y - self.current_pose.y
            distance = math.hypot(delta_x, delta_y)

            # 5. Check if goal is reached within tolerance
            if distance < 0.1:
                self.stop_turtle()
                goal_handle.succeed()
                with self.goal_lock:
                    if self.active_goal_handle == goal_handle:
                        self.active_goal_handle = None
                elapsed = (self.get_clock().now() - start_time).nanoseconds / 1e9
                self.get_logger().info(
                    f"Goal reached! Target: ({target_x:.2f}, {target_y:.2f}) | "
                    f"Distance: {total_distance:.2f}m | Time: {elapsed:.2f}s"
                )
                result.success = True
                result.total_distance = total_distance
                result.elapsed_time = elapsed
                return result

            # 5. Proportional steering and speed control
            desired_heading = math.atan2(delta_y, delta_x)
            heading_error = math.atan2(
                math.sin(desired_heading - self.current_pose.theta),
                math.cos(desired_heading - self.current_pose.theta),
            )

            cmd = Twist()
            # Clamp angular velocity within physical limits (+-2.0 rad/s)
            cmd.angular.z = max(-2.0, min(2.0, 2.5 * heading_error))
            # Advance when reasonably aligned with target
            if abs(heading_error) < 0.5:
                cmd.linear.x = min(linear_velocity, 1.5 * distance)

            # Publish to Turtlesim (/turtle1/cmd_vel)
            self.cmd_vel_pub.publish(cmd)

            # Publish to TurtleBot3 Gazebo (/cmd_vel)
            if self.use_stamped_vel:
                stamped_cmd = TwistStamped()
                stamped_cmd.header.stamp = self.get_clock().now().to_msg()
                stamped_cmd.header.frame_id = "base_footprint"
                stamped_cmd.twist = cmd
                self.tb3_cmd_vel_pub.publish(stamped_cmd)
            else:
                self.tb3_cmd_vel_pub.publish(cmd)

            # 6. Publish feedback
            feedback_msg.current_distance = distance
            feedback_msg.current_x = self.current_pose.x
            feedback_msg.current_y = self.current_pose.y
            goal_handle.publish_feedback(feedback_msg)

            loop_rate.sleep()

        result.success = False
        return result


def main(args=None):
    rclpy.init(args=args)
    node = TurtleActionServer()
    executor = MultiThreadedExecutor()
    executor.add_node(node)
    try:
        executor.spin()
    except KeyboardInterrupt:
        pass
    finally:
        node.stop_turtle()
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: 6.actions/src/action_cpp_pkg/src/turtle_action_server.cpp
#include "action_cpp_pkg/turtle_action_server.hpp"
#include "rclcpp_components/register_node_macro.hpp"

namespace action_cpp_pkg
{

TurtleActionServerNode::TurtleActionServerNode(const rclcpp::NodeOptions & options)
: Node("turtle_action_server", options)
{
    // 1. Publisher for turtle velocity commands
    cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);

    // 2. Subscriber for turtle pose
    pose_sub_ = this->create_subscription<turtlesim::msg::Pose>(
        "/turtle1/pose", 10,
        std::bind(&TurtleActionServerNode::pose_callback, this, std::placeholders::_1));

    // 3. Action server
    action_server_ = rclcpp_action::create_server<MoveToGoal>(
        this,
        "move_to_goal",
        std::bind(&TurtleActionServerNode::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
        std::bind(&TurtleActionServerNode::handle_cancel, this, std::placeholders::_1),
        std::bind(&TurtleActionServerNode::handle_accepted, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "Turtle Action Server component has been started! Action: /move_to_goal");
}

TurtleActionServerNode::~TurtleActionServerNode()
{
    stop_turtle();
}

void TurtleActionServerNode::stop_turtle()
{
    geometry_msgs::msg::Twist stop_cmd;
    cmd_vel_pub_->publish(stop_cmd);
}

rclcpp_action::GoalResponse TurtleActionServerNode::handle_goal(
    const rclcpp_action::GoalUUID & uuid,
    std::shared_ptr<const MoveToGoal::Goal> goal)
{
    (void)uuid;
    RCLCPP_INFO(this->get_logger(), "Received goal request: target=(%.2f, %.2f), velocity=%.2f",
                goal->target_x, goal->target_y, goal->linear_velocity);

    // Reject if target coordinates are outside Turtlesim canvas boundary [0.0, 11.0]
    if (goal->target_x < 0.0f || goal->target_x > 11.0f ||
        goal->target_y < 0.0f || goal->target_y > 11.0f)
    {
        RCLCPP_WARN(this->get_logger(), "Rejecting goal: target out of bounds [0.0, 11.0]!");
        return rclcpp_action::GoalResponse::REJECT;
    }

    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
}

rclcpp_action::CancelResponse TurtleActionServerNode::handle_cancel(
    const std::shared_ptr<GoalHandleMoveToGoal> goal_handle)
{
    (void)goal_handle;
    RCLCPP_INFO(this->get_logger(), "Received request to cancel goal.");
    return rclcpp_action::CancelResponse::ACCEPT;
}

void TurtleActionServerNode::handle_accepted(const std::shared_ptr<GoalHandleMoveToGoal> goal_handle)
{
    {
        std::lock_guard<std::mutex> lock(goal_mutex_);
        if (active_goal_handle_ && active_goal_handle_->is_active())
        {
            RCLCPP_INFO(this->get_logger(), "Preempting previous active goal for new target...");
        }
        active_goal_handle_ = goal_handle;
    }

    // Execute asynchronously in a separate worker thread
    std::thread{std::bind(&TurtleActionServerNode::execute, this, goal_handle)}.detach();
}

void TurtleActionServerNode::execute(const std::shared_ptr<GoalHandleMoveToGoal> goal_handle)
{
    const auto goal = goal_handle->get_goal();
    auto feedback = std::make_shared<MoveToGoal::Feedback>();
    auto result = std::make_shared<MoveToGoal::Result>();

    float target_x = goal->target_x;
    float target_y = goal->target_y;
    float linear_velocity = goal->linear_velocity > 0.0f ? goal->linear_velocity : 2.0f;

    rclcpp::Rate loop_rate(10);
    auto start_time = this->now();
    float total_distance = 0.0f;
    auto prev_pose = current_pose_;

    RCLCPP_INFO(this->get_logger(), "Executing goal: moving to (%.2f, %.2f)...", target_x, target_y);

    while (rclcpp::ok())
    {
        // 1. Check if preempted by a newer goal request
        {
            std::lock_guard<std::mutex> lock(goal_mutex_);
            if (goal_handle != active_goal_handle_)
            {
                RCLCPP_INFO(this->get_logger(), "Goal to (%.2f, %.2f) was preempted by a new goal.", target_x, target_y);
                result->success = false;
                result->total_distance = total_distance;
                result->elapsed_time = (this->now() - start_time).seconds();
                goal_handle->abort(result);
                return;
            }
        }

        // 2. Check if cancel was requested
        if (goal_handle->is_canceling())
        {
            stop_turtle();
            result->success = false;
            result->total_distance = total_distance;
            result->elapsed_time = (this->now() - start_time).seconds();
            goal_handle->canceled(result);
            RCLCPP_INFO(this->get_logger(), "Goal canceled by client.");
            {
                std::lock_guard<std::mutex> lock(goal_mutex_);
                if (active_goal_handle_ == goal_handle)
                {
                    active_goal_handle_.reset();
                }
            }
            return;
        }

        // Wait until pose is available
        if (!has_pose_)
        {
            loop_rate.sleep();
            continue;
        }

        // 3. Accumulate traveled distance
        float dx = current_pose_.x - prev_pose.x;
        float dy = current_pose_.y - prev_pose.y;
        total_distance += std::hypot(dx, dy);
        prev_pose = current_pose_;

        // 4. Calculate distance and angle to target
        float delta_x = target_x - current_pose_.x;
        float delta_y = target_y - current_pose_.y;
        float distance = std::hypot(delta_x, delta_y);

        // 5. Check if goal is reached within tolerance
        if (distance < 0.1f)
        {
            stop_turtle();
            result->success = true;
            result->total_distance = total_distance;
            result->elapsed_time = (this->now() - start_time).seconds();
            goal_handle->succeed(result);
            {
                std::lock_guard<std::mutex> lock(goal_mutex_);
                if (active_goal_handle_ == goal_handle)
                {
                    active_goal_handle_.reset();
                }
            }
            RCLCPP_INFO(this->get_logger(), "Goal reached! Target: (%.2f, %.2f) | Distance: %.2fm | Time: %.2fs",
                        target_x, target_y, total_distance, result->elapsed_time);
            return;
        }

        // 6. Proportional steering and speed control
        float desired_heading = std::atan2(delta_y, delta_x);
        float heading_error = std::atan2(
            std::sin(desired_heading - current_pose_.theta),
            std::cos(desired_heading - current_pose_.theta));

        geometry_msgs::msg::Twist cmd;
        cmd.angular.z = 4.0f * heading_error;
        if (std::abs(heading_error) < 0.5f)
        {
            cmd.linear.x = std::min(linear_velocity, 1.5f * distance);
        }
        cmd_vel_pub_->publish(cmd);

        // 7. Publish feedback
        feedback->current_distance = distance;
        feedback->current_x = current_pose_.x;
        feedback->current_y = current_pose_.y;
        goal_handle->publish_feedback(feedback);

        loop_rate.sleep();
    }

    result->success = false;
    goal_handle->abort(result);
}

void TurtleActionServerNode::pose_callback(const turtlesim::msg::Pose::SharedPtr msg)
{
    current_pose_ = *msg;
    has_pose_ = true;
}

}  // namespace action_cpp_pkg

RCLCPP_COMPONENTS_REGISTER_NODE(action_cpp_pkg::TurtleActionServerNode)
